core/src/serve_repair.rs

```diff
@@ -0,0 +1,683 @@
+    sync::{Arc, RwLock},
+    thread::{Builder, JoinHandle},
+    time::{Duration, Instant},
+};
+
+/// the number of slots to respond with when responding to `Orphan` requests
+pub const MAX_ORPHAN_REPAIR_RESPONSES: usize = 10;
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RepairType {
+    Orphan(Slot),
+    HighestShred(Slot, u64),
+    Shred(Slot, u64),
+}
+
+impl RepairType {
+    pub fn slot(&self) -> Slot {
+        match self {
+            RepairType::Orphan(slot) => *slot,
+            RepairType::HighestShred(slot, _) => *slot,
+            RepairType::Shred(slot, _) => *slot,
+        }
+    }
+}
+
+/// Window protocol messages
+#[derive(Serialize, Deserialize, Debug)]
+enum Protocol {
+    RequestWindowIndex(ContactInfo, u64, u64),
+    RequestHighestWindowIndex(ContactInfo, u64, u64),
+    RequestOrphan(ContactInfo, u64),
+}
+
+#[derive(Clone)]
+pub struct ServeRepair {
+    /// set the keypair that will be used to sign repair responses
+    keypair: Arc<Keypair>,
+    my_info: ContactInfo,
+    cluster_info: Arc<RwLock<ClusterInfo>>,
+}
+
+impl ServeRepair {
+    /// Without a valid keypair gossip will not function. Only useful for tests.
+    pub fn new_with_invalid_keypair(contact_info: ContactInfo) -> Self {
+        Self::new(Arc::new(RwLock::new(
+            ClusterInfo::new_with_invalid_keypair(contact_info.clone()),
+        )))
+    }
+
+    pub fn new(cluster_info: Arc<RwLock<ClusterInfo>>) -> Self {
+        let (keypair, my_info) = {
+            let r_cluster_info = cluster_info.read().unwrap();
+            (r_cluster_info.keypair.clone(), r_cluster_info.my_data())
+        };
+        let me = Self {
+            keypair,
+            my_info,
+            cluster_info,
+        };
+        me
+    }
+
+    pub fn my_info(&self) -> &ContactInfo {
+        &self.my_info
+    }
+
+    pub fn keypair(&self) -> &Arc<Keypair> {
+        &self.keypair
+    }
+
+    fn get_repair_sender(request: &Protocol) -> &ContactInfo {
+        match request {
+            Protocol::RequestWindowIndex(ref from, _, _) => from,
+            Protocol::RequestHighestWindowIndex(ref from, _, _) => from,
+            Protocol::RequestOrphan(ref from, _) => from,
+        }
+    }
+
+    fn handle_repair(
+        me: &Arc<RwLock<Self>>,
+        recycler: &PacketsRecycler,
+        from_addr: &SocketAddr,
+        blockstore: Option<&Arc<Blockstore>>,
+        request: Protocol,
+    ) -> Option<Packets> {
+        let now = Instant::now();
+
+        //TODO verify from is signed
+        let my_id = me.read().unwrap().keypair.pubkey();
+        let from = Self::get_repair_sender(&request);
+        if from.id == my_id {
+            warn!(
+                "{}: Ignored received repair request from ME {}",
+                my_id, from.id,
+            );
+            inc_new_counter_debug!("serve_repair-handle-repair--eq", 1);
+            return None;
+        }
+
+        /*me.write()
```

Comment:
@sagar-solana, this used to update on every repair request, I'm not sure if it's still necessary

Reply:
This just makes sure that we refresh the timestamps (liveness) of every value the from the repairee. We don't need to do it every time but we should update it at some cadence

Reply:
so it's essential repair updates this, gossip push/pull isn't enough?

Reply:
Not essential. I'm okay skipping it here.